File: problem_046.c
// -*- compile-command: "gcc -std=c99 -o problem_046 problem_046.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <stdlib.h>
#include "ceuler.h"

bool goldbach(const int composite, const int * primes, const int primeIndex)
{
    int pIndex;

    for (pIndex=0; pIndex < primeIndex; pIndex++)
    {
        int nIndex = 1;
        int conjecture = primes[pIndex] + 2 * pow(nIndex, 2);
        while (conjecture <= composite)
        {
            if (conjecture == composite) return true;
            conjecture = primes[pIndex] + 2 * pow(++nIndex, 2);
        }
    }

    return false;
}

int main()
{
    printf("Project Euler - Problem 46:\n"
           "What is the smallest odd, composite number that cannot be written as the sum of a prime and twice a square?\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int primes[10000] = {};
    int primeIndex = 0;
    int composite = 1;

    while ( composite++ )
    {
        if (isPrime(composite) == true)
            primes[primeIndex++] = composite;
        else // is composite number
        {
            if (isEven(composite) == true) continue;
            if (goldbach(composite, primes, primeIndex) == false) break;
        }
    }

    printf("Smallest composite: %d\n", composite);
    printElapsedTime(start);

    return 0;
}

File: problem_046.rs
// -*- compile-command: "rustc -o problem_046_rs problem_046.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn is_prime(n: usize) -> bool {
    if n < 2 { return false; }
    if n == 2 { return true; }
    if n % 2 == 0 { return false; }

    let limit = (n as f64).sqrt() as usize;
    for i in 3..(limit+1) {
        if n % i == 0 { return false; }
    }
    return true;
}

fn main() {
    println!("Project Euler - Problem 46:");
    println!("What is the smallest odd, composite number that cannot be written as the sum of a prime and twice a square?\n");

    let mut array = [0; 10000];
    array[2] = 1;

    for i in 2..array.len() {
        if is_prime(i) {
            let mut m = 1;
            let mut n = i + (2 * (m * m));
            while n < array.len() {
                array[n] += 1;
                m += 1;
                n = i + (2 * (m * m));
            }
        } else {
            if i % 2 != 0 && array[i] == 0 {
                println!("{}", i);
                break;
            }
        }
    }
}
